src/fsharp/FSharp.Core/tasks.fs

```diff
@@ -0,0 +1,424 @@
+                    __expand_continuation (^Awaiter : (member GetResult : unit -> ^TResult1)(awaiter))
+                else
+                    Helpers.Await __machineAddr<TaskStateMachineTemplate<'TOverall>> ((awaiter :> ICriticalNotifyCompletion), CONT)
+                    TaskStep<'TResult2, 'TOverall>(false)
+
+        static member inline GenericAwaitConfigureFalse< ^TaskLike, ^Awaitable, ^Awaiter, ^TResult1, 'TOverall
+                                                        when ^TaskLike : (member ConfigureAwait : bool -> ^Awaitable)
+                                                        and ^Awaitable : (member GetAwaiter : unit -> ^Awaiter)
+                                                        and ^Awaiter :> ICriticalNotifyCompletion 
+                                                        and ^Awaiter : (member get_IsCompleted : unit -> bool)
+                                                        and ^Awaiter : (member GetResult : unit -> ^TResult1) >
+            (task : ^TaskLike, __expand_continuation : ^TResult1 -> TaskStep<'TResult2, 'TOverall>) : TaskStep<'TResult2, 'TOverall> =
+                let awaitable = (^TaskLike : (member ConfigureAwait : bool -> ^Awaitable)(task, false))
+                TaskLikeBind<'TResult2>.GenericAwait< ^Awaitable, ^Awaiter, ^TResult1, 'TOverall>(awaitable, __expand_continuation)
+
+    /// Special case of the above for `Task<'TResult1>`. Have to write this T by hand to avoid confusing the compiler
+    /// trying to decide between satisfying the constraints with `Task` or `Task<'TResult1>`.
+    let inline bindTask (task : Task<'TResult1>) (__expand_continuation : 'TResult1 -> TaskStep<'TResult2, 'TOverall>) =
+        let CONT = __newEntryPoint()
+        let awaiter = task.GetAwaiter()
+        if awaiter.IsCompleted then 
+            __entryPoint CONT
+            __expand_continuation (awaiter.GetResult())
+        else
+            Helpers.Await __machineAddr<TaskStateMachineTemplate<'TOverall>> ((awaiter :> ICriticalNotifyCompletion), CONT)
+            TaskStep<'TResult2, 'TOverall>(false)
+
+    /// Special case of the above for `Task<'TResult1>`, for the context-insensitive builder.
+    /// Have to write this T by hand to avoid confusing the compiler thinking our built-in bind method
+    /// defined on the builder has fancy generic constraints on inp and T parameters.
+    let inline bindTaskConfigureFalse (task : Task<'TResult1>) (__expand_continuation : 'TResult1 -> TaskStep<'TResult2, 'TOverall>) =
+        let CONT = __newEntryPoint ()
+        let awaiter = task.ConfigureAwait(false).GetAwaiter()
+        if awaiter.IsCompleted then
+            __entryPoint CONT
+            __expand_continuation (awaiter.GetResult())
+        else
+            Helpers.Await __machineAddr<TaskStateMachineTemplate<'TOverall>> ((awaiter :> ICriticalNotifyCompletion), CONT)
+            TaskStep<'TResult2, 'TOverall>(false)
+
+// New style task builder.
+type TaskBuilder() =
```

Comment:
We could we make this a value type right? Otherwise sync completion is still an alloc for the TaskBuilder